caddyhttp/httpserver/middleware.go

```diff
@@ -21,7 +22,7 @@ type (
 
 	// ListenerMiddleware is similar to the Middleware type, except it
 	// chains one net.Listener to the next.
-	ListenerMiddleware func(net.Listener) net.Listener
+	ListenerMiddleware func(caddy.Listener) caddy.Listener
```

Comment:
> Are you thinking the middleware stack would live on caddy, and server types (like caddyhttp) would opt-in and use them?

Yes to the first part, and no to the second part because it's unnecessary (except for having the directive in its list of recognized directive). Remember, the HTTP server [is given a Listener as input](https://github.com/mholt/caddy/blob/55bded68c2ef23a85f643190f2444690a18f5d5d/caddyhttp/httpserver/server.go#L161) when it's time to Serve. So I think we might be able to move the listener middleware plugin functionality to the `caddy` package, because it can wrap the listeners there and then just pass the resulting Listener into Serve().

Right?

I mean, we could try it. I don't see why we would have to limit this to just the HTTP server.